pkg/operators/otel-metrics/otel-metrics.go

```diff
@@ -0,0 +1,507 @@
+	}
+
+	m.initialized = true
+
+	exporter, err := prometheus.New()
+	if err != nil {
+		return fmt.Errorf("initializing otel metrics exporter: %w", err)
+	}
+	m.exporter = exporter
+	m.meterProvider = sdkmetric.NewMeterProvider(sdkmetric.WithReader(exporter))
+
+	if !m.skipListen {
+		go func() {
+			mux := http.NewServeMux()
+			mux.Handle("/metrics", promhttp.Handler())
+			err := http.ListenAndServe(globalParams.Get(ParamOtelMetricsListenAddress).AsString(), mux)
+			if err != nil {
+				log.Errorf("serving otel metrics on: %s", err)
+				return
+			}
+		}()
+	}
+	return nil
+}
+
+func (m *otelMetricsOperator) GlobalParams() api.Params {
+	return api.Params{
+		{
+			Key:          ParamOtelMetricsEnabled,
+			DefaultValue: "false",
+			TypeHint:     api.TypeBool,
+		},
+		{
+			Key:          ParamOtelMetricsListenAddress,
+			DefaultValue: "0.0.0.0:2224",
+			TypeHint:     api.TypeString,
+		},
+	}
+}
+
+func (m *otelMetricsOperator) InstanceParams() api.Params {
+	return api.Params{
+		{
+			Key:         ParamOtelMetricsName,
+			TypeHint:    api.TypeString,
+			Description: "override name of the exported datasource; use a comma-separated list with datasource:newname to specify more than one name",
+		},
+	}
+}
+
+func (m *otelMetricsOperator) InstantiateDataOperator(gadgetCtx operators.GadgetContext, instanceParamValues api.ParamValues) (operators.DataOperatorInstance, error) {
+	if !m.initialized {
+		return nil, nil
+	}
+
+	// extract name mappings; key will be old name (or empty), value the new name
+	mappings := make(map[string]string)
+	for _, m := range strings.Split(instanceParamValues[ParamOtelMetricsName], ",") {
+		names := strings.SplitN(m, ":", 2)
+		from := ""
+		to := names[0]
+		if len(names) == 2 {
+			from = to
+			to = names[1]
+		}
+		mappings[from] = to
+	}
+
+	instance := &otelMetricsOperatorInstance{
+		op:           m,
+		collectors:   make(map[datasource.DataSource]*metricsCollector),
+		nameMappings: mappings,
+	}
+
+	err := instance.init(gadgetCtx)
+	if err != nil {
+		return nil, err
+	}
+	return instance, nil
+}
+
+func (m *otelMetricsOperator) Priority() int {
+	return Priority
+}
+
+type otelMetricsOperatorInstance struct {
+	op           *otelMetricsOperator
+	collectors   map[datasource.DataSource]*metricsCollector
+	nameMappings map[string]string
+}
+
+func (m *otelMetricsOperatorInstance) Name() string {
+	return name
+}
+
+type metricsCollector struct {
+	meter  metric.Meter
+	keys   []func(datasource.Data) attribute.KeyValue
+	values []func(context.Context, datasource.Data, attribute.Set)
+}
+
+func asInt64Func[T constraints.Integer](extract func(datasource.Data) (T, error)) func(datasource.Data) int64 {
+	return func(data datasource.Data) int64 {
+		v, _ := extract(data)
+		return int64(v)
+	}
+}
+
+func asInt64(f datasource.FieldAccessor) func(datasource.Data) int64 {
+	switch f.Type() {
+	default:
+		// This should not be called with types other than int
+		panic("unsupported field type for asInt64")
```

Comment:
It's a code path that shouldn't be used (validation should happen (and happens) in the callers of this function) - that's why I opted not to have runtime error handling here. Should this ever be used, it will let us know immediately.

Reply:
I think it's better to have these "asserts" in (that would crash and instantly notify us), rather than just returning with an unexpected 0 in those cases (should anyone ever use it in a wrong way.)